Linker error calling certain functions

The metadata indicates the name of the DLL that exports the given function, but Rust doesn't yet support DLL imports and expects a LIB file to resolve those imports. If you have the Windows SDK installed, there will usually be a LIB file by the right name that Rust can use but this is not guaranteed. There are a few edge cases where this doesn't quite line up. I considered generating [my own LIB files](https://github.com/microsoft/windows-rs/issues/142) for the `windows` crate but we have some folks looking into implementing [DLL imports](https://github.com/rust-lang/rfcs/blob/master/text/2627-raw-dylib-kind.md) for Rust and I'm hopeful that will arrive soon. 

One workaround is to generate and provide your own lib files. This process is [briefly described here](https://github.com/microsoft/windows-rs/issues/142#issuecomment-741054934).

Another workaround is to copy the declaration just for the function in question and change the name of the lib to match a suitable lib that does support the function. The msdn docs usually have the correct name of a lib.

I ran into linker issues when trying to use the [HasExpandedResources](https://docs.microsoft.com/en-us/windows/win32/api/expandedresources/nf-expandedresources-hasexpandedresources) API. Here's my code:

```
mod bindings {
    ::windows::include_bindings!();
}

use bindings::{
    windows::win32::game_mode::*,
};

fn main() {
    unsafe {
        let mut game_mode: i32 = 0;
        let error = HasExpandedResources(&mut game_mode);
        if error.is_ok() {
            println!("Running in game mode: {}", game_mode != 0);
        }
    }
```
Linker says:
```
...
= note: LINK : fatal error LNK1181: cannot open input file 'api-ms-win-gaming-expandedresources-l1-1-0.lib'
```

The docs for this function say to link against WindowsApp.lib though, so not sure why that API set ends up here...

--- a/crates/gen/src/function.rs
+++ b/crates/gen/src/function.rs
@@ -31,12 +31,10 @@ impl Function {
             TokenStream::new()
         };
 
-        // TODO: need to generate libs until Rust supports dynamic linking against DLLs.
-        // This is actually the DLL name.
         let mut link = self.signature.method.impl_map().unwrap().scope().name();
-        if link == "ext-ms-win-core-iuri-l1-1-0" {
-            link = "urlmon";
-        } else if link.starts_with("api-ms-win-core") {
+
+        // TODO: workaround for https://github.com/microsoft/windows-rs/issues/463
+        if link.contains("-ms-win-") || link == "D3DCOMPILER_47" {
             link = "onecoreuap";
         }
 

--- a/crates/tests/build.rs
+++ b/crates/tests/build.rs
@@ -63,10 +63,13 @@ fn main() {
         windows::win32::windows_programming::{
             CloseHandle
         },
-        windows::win32::com::{IUri, CreateUri},
+        windows::win32::com::CreateUri,
         windows::win32::structured_storage::{CreateStreamOnHGlobal, STREAM_SEEK},
         windows::win32::upnp::UIAnimationTransitionLibrary,
         windows::win32::ldap::ldapsearch,
         windows::win32::upnp::UIAnimationManager,
+        windows::win32::game_mode::HasExpandedResources,
+        windows::win32::debug::MiniDumpWriteDump,
+        windows::win32::direct3d11::D3DDisassemble11Trace,
     );
 }

--- a/crates/tests/tests/win32.rs
+++ b/crates/tests/tests/win32.rs
@@ -1,6 +1,9 @@
 use windows::Abi;
 
 use tests::{
+    windows::win32::com::CreateUri,
+    windows::win32::debug::{MiniDumpWriteDump, MINIDUMP_TYPE},
+    windows::win32::direct3d11::D3DDisassemble11Trace,
     windows::win32::direct3d12::D3D12_DEFAULT_BLEND_FACTOR_ALPHA,
     windows::win32::direct3d_hlsl::D3DCOMPILER_DLL,
     windows::win32::display_devices::RECT,
@@ -8,6 +11,7 @@ use tests::{
         CreateDXGIFactory1, IDXGIFactory7, DXGI_ADAPTER_FLAG, DXGI_FORMAT, DXGI_MODE_DESC,
         DXGI_MODE_SCALING, DXGI_MODE_SCANLINE_ORDER, DXGI_RATIONAL,
     },
+    windows::win32::game_mode::HasExpandedResources,
     windows::win32::ldap::ldapsearch,
     windows::win32::security::ACCESS_MODE,
     windows::win32::structured_storage::{CreateStreamOnHGlobal, STREAM_SEEK},
@@ -243,6 +247,45 @@ fn com_inheritance() {
                 .0
                 == DXGI_ERROR_INVALID_CALL as u32
         );
+    }
+}
+
+// Tests for https://github.com/microsoft/windows-rs/issues/463
+#[test]
+fn onecore_imports() -> windows::Result<()> {
+    unsafe {
+        let mut has_expanded_resources = 0;
+        HasExpandedResources(&mut has_expanded_resources).ok()?;
+
+        let mut uri = None;
+        let uri = CreateUri(
+            windows::HString::from("http://kennykerr.ca")
+                .as_wide()
+                .as_ptr(),
+            0,
+            0,
+            &mut uri,
+        )
+        .and_some(uri)?;
+
+        let mut port = 0;
+        uri.GetPort(&mut port).ok()?;
+        assert!(port == 80);
+
+        let result = MiniDumpWriteDump(
+            HANDLE(0),
+            0,
+            HANDLE(0),
+            MINIDUMP_TYPE::MiniDumpNormal,
+            std::ptr::null_mut(),
+            std::ptr::null_mut(),
+            std::ptr::null_mut(),
+        );
+        assert!(result.is_err());
+
+        assert!(D3DDisassemble11Trace(std::ptr::null_mut(), 0, None, 0, 0, 0, &mut None).is_err());
+
+        Ok(())
     }
 }
 

